pkg/operation/botanist/health_check.go

```diff
@@ -414,40 +398,34 @@ func (b *Botanist) checkAlerts(checker *HealthChecker, condition gardencorev1bet
 	return gardencorev1beta1helper.UpdatedCondition(condition, gardencorev1beta1.ConditionTrue, reason, message.String())
 }
 
-// CheckClusterNodes checks whether cluster nodes in the given listers are complete and healthy.
+// CheckClusterNodes checks whether cluster nodes in the given listers are healthy and within the desired range.
+// Additional checks are executed in the provider extension
 func (b *HealthChecker) CheckClusterNodes(
-	namespace string,
+	workerPools []gardencorev1beta1.Worker,
 	condition gardencorev1beta1.Condition,
 	nodeLister kutil.NodeLister,
-	machineDeploymentLister kutil.MachineDeploymentLister,
 ) (*gardencorev1beta1.Condition, error) {
-	nodeList, err := nodeLister.List(labels.Everything())
-	if err != nil {
-		return nil, err
-	}
-
-	if exitCondition := b.checkNodes(condition, nodeList); exitCondition != nil {
-		return exitCondition, nil
-	}
-
-	machineDeploymentList, err := machineDeploymentLister.MachineDeployments(namespace).List(labels.Everything())
-	if err != nil {
-		return nil, err
-	}
+	for _, worker := range workerPools {
+		requirement, err := labels.NewRequirement("worker.gardener.cloud/pool", selection.Equals, []string{worker.Name})
+		if err != nil {
+			return nil, err
+		}
+		nodeList, err := nodeLister.List(labels.NewSelector().Add(*requirement))
+		if err != nil {
+			return nil, err
+		}
 
-	desiredMachines := 0
-	for _, machineDeployment := range machineDeploymentList {
-		if machineDeployment.DeletionTimestamp == nil {
-			desiredMachines += int(machineDeployment.Spec.Replicas)
+		if exitCondition := b.checkNodes(condition, nodeList, worker.Name); exitCondition != nil {
```

Comment:
Why do we do need to do these checks? They are already done by the worker health controller, right?

Reply:
The tests have different scopes
- gardenlet: check that each worker pool has sufficient nodes available (> min , <  max).
It only requires connectivity to the shoot.
- extension: check that machine deployment is healthy && that as many nodes are registered to the cluster as machines are created (all nodes joined the cluster). Requires connectivity to seed and shoot.

I think it would make sense that the gardenlet still performs this basic check. In case it cannot connect to the seed cluster to collect the extension health information from the worker, it should still be able to verify the shoot. 
Plus they have slightly different scopes.
Some redundancy for safety is also not bad IMHO

Reply:
OK, thanks for clarifying, agreed. 👍